full-stack-intro/client/src/components/Todos.tsx

```diff
@@ -0,0 +1,156 @@
+import { useEffect, useState } from 'react';
+import { PageTitle } from './PageTitle';
+import { TodoList } from './TodoList';
+import { TodoForm } from './TodoForm';
+
+export type UnsavedTodo = {
+  task: string;
+  isCompleted: boolean;
+};
+export type Todo = UnsavedTodo & {
+  todoId: number;
+};
+
+export function Todos() {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [isLoading, setIsLoading] = useState(true);
+  const [error, setError] = useState<unknown>();
+
+  /* Implement useEffect to fetch all todos. Hints are at the bottom of the file. */
+  useEffect(() => {
+    async function read() {
+      try {
+        const response = await fetch('/api/todos');
+        if (!response.ok) throw new Error(`Error: ${response.status}`);
+
+        const todoRead = await response.json();
+        setTodos(todoRead);
+      } catch (err) {
+        setError(err);
+      } finally {
+        setIsLoading(false);
+      }
+    }
+
+    read();
+  }, []);
+
+  /* Implement addTodo to add a new todo. Hints are at the bottom of the file. */
+  async function addTodo(newTodo: UnsavedTodo) {
+    try {
+      const response = await fetch('/api/todos', {
+        method: 'POST',
+        body: JSON.stringify(newTodo),
+        headers: { 'Content-Type': 'application/json' },
+      });
+      if (!response.ok) throw new Error(`Error: ${response.status}`);
+      const addedTodo = await response.json();
+      const updateTodos = structuredClone(todos);
```

Comment:
I wasn't aware of `structuredClone`! I had to read up on it. Since it's new, it may not be in all browsers, but it's likely in all the ones you care about. It also makes a deep copy (which means it copies all the objects within objects, recursively). That's more expensive than you need. For React state setter purposes, a shallow copy is fine. But there's no harm in a deep copy (at least, none I can think of right now).

I'm used to just using:
`const updateTodos = [...todos, addedTodo];`